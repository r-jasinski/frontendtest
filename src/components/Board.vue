<template>
  <div class="board" @keydown.prevent tabindex="0">
    <span
      v-for="i in Array.from({ length: 64 }, (_, i) => 64 - i)"
      :data-file="getDataFile(i)"
      :data-rank="getDataRank(i)"
      :key="i"
      @click="onSquareSelect(i)"
      tabindex="1"
    >
    </span>
    <GearsSVG />
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'Board',
  props: { onSquareSelect: { type: Function, required: true } },
  methods: {
    getDataRank(square: number) {
      const rank = Math.ceil(square / 8)
      return square % 8 === 0 ? rank : null
    },
    getDataFile(square: number) {
      return square <= 8 ? String.fromCharCode(104 - ((square - 1) % 8)) : null
    }
  }
})
</script>

<style scoped>
.board {
  height: 100%;
  aspect-ratio: 1;
  position: relative;
  display: grid;
  grid-template-columns: repeat(8, 1fr);
  clip-path: inset(0 0 0 0 round 4px);
  background-color: var(--color-gray-300);
  color: var(--color-green-300);
}
span:nth-child(1) ~ span:nth-child(-2n + 8),
span:nth-child(8) ~ span:nth-child(-2n + 15),
span:nth-child(16) ~ span:nth-child(-2n + 24),
span:nth-child(24) ~ span:nth-child(-2n + 31),
span:nth-child(32) ~ span:nth-child(-2n + 40),
span:nth-child(40) ~ span:nth-child(-2n + 47),
span:nth-child(48) ~ span:nth-child(-2n + 56),
span:nth-child(56) ~ span:nth-child(-2n + 65) {
  background-color: var(--color-green-300);
  color: var(--color-gray-300);
}
span {
  width: 100%;
  height: 100%;
  aspect-ratio: 1;
  position: relative;
  outline: none;
  border: none;
  font-size: 12px;
  font-weight: bold;
  cursor: pointer;
}
span:focus:after {
  content: '';
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: rgb(255, 255, 51);
  opacity: 0.5;
}
span[data-rank]:focus:after {
  width: calc(100% - 4px);
  height: calc(100% - 4px);
}
span[data-rank]:after {
  content: attr(data-rank);
  position: absolute;
  padding-left: 4px;
  padding-top: 4px;
}
span[data-file]:before {
  content: attr(data-file);
  position: absolute;
  bottom: 0;
  right: 0;
  margin: 4px;
}
</style>
