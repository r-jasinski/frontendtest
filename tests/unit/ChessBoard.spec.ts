import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import Board from '../../src/components/ChessBoard.vue'

describe('Board.vue', () => {
  const wrapper = mount(Board)
  it('should render board squares', () => {
    expect(wrapper.findAll('.board span')).toHaveLength(64)
  })
  it('should render board rank legend', () => {
    const ranks = ['1', '2', '3', '4', '5', '6', '7', '8']
    ranks.forEach((rank) => {
      expect(wrapper.find(`[data-rank='${rank}']`).exists()).toBeTruthy()
    })
  })
  it('should render board file legend', () => {
    const files = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']
    files.forEach((file) => {
      expect(wrapper.find(`[data-file='${file}']`).exists()).toBeTruthy()
    })
  })
  it('emit event and value when squared is clicked', async () => {
    await wrapper.find("[data-file='a'][data-rank='1']").trigger('click')
    expect(wrapper.emitted('square-select')?.[0]?.[0]).toBe(8)
  })
  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
