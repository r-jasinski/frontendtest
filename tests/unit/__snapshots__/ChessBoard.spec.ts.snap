// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Board.vue > should match snapshot 1`] = `
<div
  class="board"
  data-v-41886935=""
  tabindex="0"
>
  
  <span
    data-rank="8"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-rank="7"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-rank="6"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-rank="5"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-rank="4"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-rank="3"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-rank="2"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="a"
    data-rank="1"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="b"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="c"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="d"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="e"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="f"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="g"
    data-v-41886935=""
    tabindex="1"
  />
  <span
    data-file="h"
    data-v-41886935=""
    tabindex="1"
  />
  
</div>
`;
